Dnn detect tensorflow support

Semmer2 opened the pull request:
Support dnn detect only, do not support classify for now

guoyejun committed:
lavfi: show side data of bounding box

Signed-off-by: Guo, Yejun <yejun.guo@intel.com>

guoyejun committed:
lavfi: add filter dnn_detect for object detection

Below are the example steps to do object detection:

1. download and install l_openvino_toolkit_p_2021.1.110.tgz from
https://software.intel.com/content/www/us/en/develop/tools/openvino-toolkit/download.html
  or, we can get source code (tag 2021.1), build and install.
2. export LD_LIBRARY_PATH with openvino settings, for example:
.../deployment_tools/inference_engine/lib/intel64/:.../deployment_tools/inference_engine/external/tbb/lib/
3. rebuild ffmpeg from source code with configure option:
--enable-libopenvino
--extra-cflags='-I.../deployment_tools/inference_engine/include/'
--extra-ldflags='-L.../deployment_tools/inference_engine/lib/intel64'
4. download model files and test image
wget https://github.com/guoyejun/ffmpeg_dnn/raw/main/models/openvino/2021.1/face-detection-adas-0001.bin
wget https://github.com/guoyejun/ffmpeg_dnn/raw/main/models/openvino/2021.1/face-detection-adas-0001.xml
wget
https://github.com/guoyejun/ffmpeg_dnn/raw/main/models/openvino/2021.1/face-detection-adas-0001.label
wget https://github.com/guoyejun/ffmpeg_dnn/raw/main/images/cici.jpg
5. run ffmpeg with:
./ffmpeg -i cici.jpg -vf dnn_detect=dnn_backend=openvino:model=face-detection-adas-0001.xml:input=data:output=detection_out:confidence=0.6:labels=face-detection-adas-0001.label,showinfo -f null -

We'll see the detect result as below:
[Parsed_showinfo_1 @ 0x55978db02dc0]   side data - bounding boxes:
[Parsed_showinfo_1 @ 0x55978db02dc0] source: face-detection-adas-0001.xml
[Parsed_showinfo_1 @ 0x55978db02dc0] index: 0, region: (1005, 813) -> (1086, 905), label: face, confidence: 10000/10000.
[Parsed_showinfo_1 @ 0x55978db02dc0] index: 1, region: (888, 839) -> (967, 926), label: face, confidence: 6917/10000.

There are two faces detected with confidence 100% and 69.17%.

Signed-off-by: Guo, Yejun <yejun.guo@intel.com>

guoyejun committed:
lavu: add side data AV_FRAME_DATA_BOUNDING_BOXES

Signed-off-by: Guo, Yejun <yejun.guo@intel.com>

guoyejun committed:
lavfi/dnn: refine code for frame pre/proc processing

Semmer2 committed:
lavfi/dnn_backend_tensorflow.c: support detect model as input model

This patch enables tensorflow backend supporting detect model as
inference model. There exist 4 output nodes in current supported
detect model. But this patch did not fullfill all the code tensorflow
detect model inference need.

guoyejun committed:
lavfi/dnn: add post process for detection

guoyejun committed:
lavfi/dnn_backend_openvino.c: only allow DFT_PROCESS_FRAME to get output dim

Semmer2 committed:
dnn/vf_dnn_detect: add tensorflow output parse support

Testing model is tensorflow offical model in github repo, please refer
https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf1_detection_zoo.md
to download the detect model as you need.
For examplei, local testing was carried on with 'ssd_mobilenet_v1_coco', and
used one image of dog in
https://github.com/tensorflow/models/blob/master/research/object_detection/test_images/image1.jpg

Testing command is:
./ffmpeg -i image1.jpg -vf dnn_detect=dnn_backend=tensorflow:input=image_tensor:\
output="num_detections&detection_scores&detection_classes&detection_boxes":\
model=ssd_mobilenet_v1_coco.pb:async=0,showinfo -f null -

We will see the result similar as below:
[Parsed_showinfo_1 @ 0x4e6d800]   side data - bounding boxes:
[Parsed_showinfo_1 @ 0x4e6d800] source: ssd_mobilenet_v1_coco.pb
[Parsed_showinfo_1 @ 0x4e6d800] index: 0,       region: (19, 24) -> (323, 554), label: 18, confidence: 9405/10000.
[Parsed_showinfo_1 @ 0x4e6d800] index: 1,       region: (412, 69) -> (996, 587), label: 18, confidence: 9345/10000.

There are two dogs detected with core 94.05% and 93.45%.

guoyejun commented on libavfilter/dnn/dnn_backend_tf.c:
```diff
@@ -804,6 +805,21 @@ static DNNReturnType execute_model_tf(const DNNModel *model, const char *input_n
+            av_log(ctx, AV_LOG_ERROR, "detect filter needs to provide post proc\n");
+            return DNN_ERROR;
+        }
+        model->detect_post_proc(out_frame, output, nb_output, model->filter_ctx);
```
does your detection work without this patch?

Semmer2 replied:
Yes, I tested tf detection with a picture of dog, and it works well with all my patches applied.

guoyejun replied:
we need to merge part of this change into others, and in the last patch to enable tf+detect, we'd also add detail command lines.

Semmer2 replied:
OK, sure

guoyejun commented on libavfilter/dnn/dnn_backend_tf.c:
```diff
@@ -153,7 +153,7 @@ static DNNReturnType get_input_tf(void *model, DNNData *input, const char *input
 
     // currently only NHWC is supported
-    av_assert0(dims[0] == 1);
+    av_assert0(dims[0] == 1 || dims[0] == -1);
```
is this change enough to "support model with random input"?

btw, the title 'random input' is not clear.

Semmer2 replied:
Just for fixing the error of "input dims[0] != 1". Maybe i should change this commit message to something like: dnn/dnn_backend_tf: fix error of model input batch_size = -1 ?

guoyejun replied:
dims[0] is batch size, we'd put all the changes relative to batch size together as a single patch

Semmer2 replied:
OK

guoyejun commented on libavfilter/dnn/dnn_io_proc.c:
```diff
@@ -168,8 +168,12 @@ static DNNReturnType proc_from_frame_to_dnn_frameprocessing(AVFrame *frame, DNND
+            return AV_PIX_FMT_BGR24;
+        } else if (data->order == DCO_RGB) {
+            return AV_PIX_FMT_RGB24;
+        }
```
use switch/case, not if/else

Semmer2 replied:
OK

guoyejun commented on libavfilter/dnn/dnn_backend_tf.c:
```diff
@@ -790,22 +784,25 @@ static DNNReturnType execute_model_tf(const DNNModel *model, const char *input_n
+                tf_model->model->frame_post_proc(out_frame, output, tf_model->model->filter_ctx);
             } else {
-                ff_proc_from_dnn_to_frame(out_frame, &output, ctx);
+                ff_proc_from_dnn_to_frame(out_frame, &output[0], ctx);
```
callback for detection missed

Semmer2 replied:
Yes, this patch does not work well, detect callback func is in the last patch. Maybe current patch set is not resonable, i will think again.

guoyejun replied:
yes, thanks.

guoyejun commented on libavfilter/dnn_filter_common.h:
```diff
@@ -51,6 +53,7 @@ int ff_dnn_init(DnnContext *ctx, DNNFunctionType func_type, AVFilterContext *fil
 int ff_dnn_set_frame_proc(DnnContext *ctx, FramePrePostProc pre_proc, FramePrePostProc post_proc);
 int ff_dnn_set_detect_post_proc(DnnContext *ctx, DetectPostProc post_proc);
 int ff_dnn_set_classify_post_proc(DnnContext *ctx, ClassifyPostProc post_proc);
+char **ff_dnn_parse_values(DnnContext *ctx, const char *output_exp, const char *val_sep, int *nb_output);
```
what does this function do?  We need a meaningful function name.

Semmer2 replied:
To parse multiple outputs, with '&' to seperate. Maybe parse_output_name_expr better?
While, i thought may be there would be other condition, maybe this func would be called to parse multiple names/expressions with some certain seperation symbol. So i choosed a more common func name.

guoyejun replied:
we still need a meaningful name even it is common, for example, separate_expressions or something else.

Semmer2 replied:
Yes, i get it

guoyejun commented on libavfilter/dnn_filter_common.c:
```diff
@@ -19,6 +19,44 @@
+    case DFT_ANALYTICS_DETECT:
+        switch(backend_type) {
+        case DNN_TF:
+            if (output_nb != 4) {
```
this logic is special in vf_detect.c for tf backend, why we check here?

Semmer2 replied:
I think maybe tf classify model may reuse this logic code in following code. Since tf calssify model needs multiple outputs as well.

guoyejun replied:
the magic number 4 should be special

Semmer2 replied:
OK~ So is it ok that i put the logic here?

guoyejun replied:
just put in vf_dnn_detect.c now, we can refine the code if it is common later

Semmer2 replied:
OK, got you

guoyejun commented on libavfilter/vf_dnn_detect.c:
```diff
@@ -62,29 +62,30 @@ static const AVOption dnn_detect_options[] = {
+                                        const float *conf_p,         const float *label_id_p,
+                                        const float *position_p,     const int proposal_count,
+                                        const int x0_offset,         const int y0_offset,
+                                        const int x1_offset,         const int y1_offset)
```
parse_output_to_bounding_box is designed as a general function, but it would be very complex when more models are supported, for example, some model might not output confidence, etc.

so, we can rename the original dnn_detect_post_proc to dnn_detect_post_proc_ov, and add a new function dnn_detect_post_proc_tf. Keep them separated.

Semmer2 replied:
Oh, you're right, but these two func have so many common codes, maybe I should use a macro? Or just leave it?

guoyejun replied:
we can just leave it unless there's other comment

Semmer2 replied:
OK

Semmer2 committed:
dnn: add DCO_RGB color order to enum DNNColorOrder

Adding DCO_RGB color order to DNNColorOrder, since tensorflow model
needs this kind of color oder as input.

guoyejun commented on libavfilter/vf_dnn_detect.c:
```diff
@@ -211,7 +214,12 @@ static int read_detect_label_file(AVFilterContext *context)
+    DnnContext *dnn_ctx = &ctx->dnnctx;
+    int ret;
+
+    dnn_ctx->model_outputname = ff_dnn_separate_expression(dnn_ctx, dnn_ctx->outputname_expr,
```
better to be done within ff_dnn_init, in other words, don't export function ff_dnn_separate_expression

guoyejun commented on libavfilter/dnn_filter_common.h:
```diff
@@ -30,7 +30,9 @@ typedef struct DnnContext {
     DNNBackendType backend_type;
     char *model_inputname;
-    char *model_outputname;
+    char **model_outputname;
```
model_output_names

guoyejun commented on libavfilter/vf_dnn_detect.c:
```diff
@@ -48,6 +48,9 @@ typedef struct DnnDetectContext {
     { "dnn_backend", "DNN backend",                OFFSET(backend_type),     AV_OPT_TYPE_INT,       { .i64 = 2 },    INT_MIN, INT_MAX, FLAGS, "backend" },
+#if (CONFIG_LIBTENSORFLOW == 1)
+    { "tensorflow",  "tensorflow backend flag",    0,                        AV_OPT_TYPE_CONST,     { .i64 = 1 },    0, 0, FLAGS, "backend" },
+#endif
```
if you enable this option in this patch, the tf backend should work here

guoyejun commented on libavfilter/vf_dnn_processing.c:
```diff
@@ -62,6 +62,9 @@ AVFILTER_DEFINE_CLASS(dnn_processing);
     DnnProcessingContext *ctx = context->priv;
+    DnnContext *dnn_ctx = &ctx->dnnctx;
+    dnn_ctx->model_outputname = ff_dnn_separate_expression(dnn_ctx, dnn_ctx->outputname_expr,
+                                                           "&", &dnn_ctx->nb_output);
```
this patch shows that: dnn_processing is broken by your previous patch, it is not allowed. Please think carefully about how to split/merge the patches.

guoyejun commented on libavfilter/vf_dnn_detect.c:
```diff
@@ -211,6 +211,30 @@ static int read_detect_label_file(AVFilterContext *context)
+    switch(backend_type) {
+    case DNN_TF:
+        if (output_nb != 4) {
+            av_log(ctx, AV_LOG_ERROR, "Dnn detect filter with tensorflow backend needs 4 outputs, \
```
it is not tf backend needs 4 outputs, but that we currently only support models with 4 outputs.

guoyejun commented on libavfilter/vf_dnn_detect.c:
```diff
@@ -62,9 +62,10 @@ static const AVOption dnn_detect_options[] = {
+static int dnn_detect_post_proc_ov(AVFrame *frame, DNNData *output, uint32_t nb, AVFilterContext *filter_ctx)
 {
     DnnDetectContext *ctx = filter_ctx->priv;
+    float x0, y0, x1, y1;
```
don't change the code that is not relative to your patch

guoyejun commented on libavfilter/vf_dnn_detect.c:
```diff
@@ -95,6 +96,7 @@ static int dnn_detect_post_proc(AVFrame *frame, DNNData *output, uint32_t nb, AV
 
     sd = av_frame_new_side_data(frame, AV_FRAME_DATA_BOUNDING_BOXES,
                                 sizeof(*header) + sizeof(*bbox) * nb_bbox);
+
```
don't change the code that is not relative to your patch

guoyejun commented on libavfilter/vf_dnn_detect.c:
```diff
@@ -249,7 +332,17 @@ static av_cold int dnn_detect_init(AVFilterContext *context)
     ret = ff_dnn_init(&ctx->dnnctx, DFT_ANALYTICS_DETECT, context);
     if (ret < 0)
         return ret;
-    ff_dnn_set_detect_post_proc(&ctx->dnnctx, dnn_detect_post_proc);
```
still use dnn_detect_post_proc as a single entry, but to go to different path within this function.

guoyejun commented on libavfilter/vf_drawbox.c:
```diff
@@ -389,58 +397,8 @@ static av_pure av_always_inline int pixel_belongs_to_grid(DrawBoxContext *drawgr
-            }
-        }
-    }
+    draw_box_and_grid(frame, drawgrid, 0, 0, frame->width, frame->height, pixel_belongs_to_grid);
```
please just remove the patches about vf_drawbox/text in this patch set, let's only focus on detect+tf here

guoyejun commented:
basically, you can merge them into two patches, one for the tf backend, the other for the detect filter.  And 1-2 more small patches is possible if you think some changes in tf backend is not special for detect.

guoyejun commented:
btw, please rebase your patches on my detect patch (do not include classify), so your patches can be merged after my detect is pushed.

Semmer2 commented:
OK, got yout
